### PR TITLE
porous flow mass and energy conservation fixes.

### PR DESCRIPTION
**Pull request by WilkAndy**
Refs #18324

### CODE REVIEW

**Comment by WilkAndy** [comment]
Hi @cpgr - i'm wondering whether you can review this, please?  You'll see the code changes in the first commit, and they're quite minor, but necessitate a large number of input file changes and minor regolding.  The doc is close to a copy-and-paste the the file i sent you a little while ago.

File: modules/porous_flow/src/actions/PorousFlowActionBase.C [hunk 1]
```diff
@@ -453,7 +453,8 @@ PorousFlowActionBase::addVolumetricStrainMaterial(const std::vector<VariableName
     std::string material_name = "PorousFlowActionBase_VolumetricStrain";
     params.set<UserObjectName>("PorousFlowDictator") = _dictator_name;
     params.set<std::vector<VariableName>>("displacements") = displacements;
-    params.set<bool>("consistent_with_displaced_mesh") = consistent_with_displaced_mesh;
+    if (base_name != "")
```
Comment (cpgr) [inline modules/porous_flow/src/actions/PorousFlowActionBase.C:456]: Minor, but I like `if (!base_name.empty())`

File: modules/porous_flow/doc/content/source/kernels/PorousFlowHeatVolumetricExpansion.md [hunk 2]
```diff
@@ -8,6 +8,9 @@ This `Kernel` implements the weak form of
 \end{equation*}
 where all parameters are defined in the [nomenclature](/nomenclature.md).
 
+!alert note
+This Kernel should employ `use_displaced_mesh = false`.  Further information can be found [here](porous_flow/time_derivative.md)
+
```
[the remark this answers is not in the document]
Comment (cpgr) [reply modules/porous_flow/doc/content/source/kernels/PorousFlowHeatVolumetricExpansion.md:13]: Oh, I see that you have done this once I got to the code.

File: modules/porous_flow/src/actions/PorousFlowFullySaturated.C [hunk 3]
```diff
@@ -172,6 +172,8 @@ PorousFlowFullySaturated::addKernels()
     params.set<UserObjectName>("PorousFlowDictator") = _dictator_name;
     params.set<bool>("multiply_by_density") = _multiply_by_density;
     params.set<bool>("strain_at_nearest_qp") = _strain_at_nearest_qp;
+    if (_base_name != "")
+      params.set<std::string>("base_name") = _base_name;
```
Comment (cpgr) [inline modules/porous_flow/src/actions/PorousFlowFullySaturated.C:176]: Maybe `empty()`

File: modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C [hunk 4]
```diff
@@ -263,6 +267,8 @@ PorousFlowSinglePhaseBase::addKernels()
       params.set<NonlinearVariableName>("variable") = _temperature_var[0];
       params.set<UserObjectName>("PorousFlowDictator") = _dictator_name;
       params.set<bool>("strain_at_nearest_qp") = _strain_at_nearest_qp;
+      if (_base_name != "")
```
Comment (cpgr) [inline modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C:270]: maybe `empty()`

File: modules/porous_flow/src/actions/PorousFlowUnsaturated.C [hunk 5]
```diff
@@ -161,6 +161,8 @@ PorousFlowUnsaturated::addKernels()
     InputParameters params = _factory.getValidParams(kernel_type);
     params.set<UserObjectName>("PorousFlowDictator") = _dictator_name;
     params.set<bool>("strain_at_nearest_qp") = _strain_at_nearest_qp;
+    if (_base_name != "")
```
Comment (cpgr) [inline modules/porous_flow/src/actions/PorousFlowUnsaturated.C:164]: maybe `empty()`

File: modules/porous_flow/test/tests/energy_conservation/tests [hunk 6]
```diff
@@ -47,6 +47,16 @@
...
+  [./heat03_rz]
+    type = 'CSVDiff'
+    input = 'heat03_rz.i'
+    csvdiff = 'heat03_rz_csv.csv'
+    rel_err = 1.0E-5
+    threading = '!pthreads'
+    issues = '#18324'
+    design = 'PorousFlowHeatEnergy.md PorousFlowHeatVolumetricExpansion.md PorousFlowEnergyTimeDerivative.md PorousFlowFullySaturatedHeatAdvection.md PorousFlowHeatAdvection.md PorousFlowHeatConduction.md PorousFlowPlasticHeatEnergy.md'
+    requirement = 'PorousFlow shall correctly conserve heat energy in a THM simulation when the model is squashed mechanically, in RZ coordinates'
+  [../]
```
Comment (cpgr) [inline modules/porous_flow/test/tests/energy_conservation/tests:59]: Hehehe - squashed!

File: modules/porous_flow/test/tests/mass_conservation/mass12.i [hunk 7]
```diff
@@ -0,0 +1,283 @@
+# The sample is a annulus in RZ coordinates.
```
Comment (cpgr) [inline modules/porous_flow/test/tests/mass_conservation/mass12.i:1]:
```suggestion
# The sample is an annulus in RZ coordinates.
```

File: modules/porous_flow/test/tests/mass_conservation/mass13.i [hunk 8]
```diff
@@ -0,0 +1,368 @@
+# The sample is a annulus in RZ coordinates.
```
Comment (cpgr) [inline modules/porous_flow/test/tests/mass_conservation/mass13.i:1]:
```suggestion
# The sample is an annulus in RZ coordinates.
```

**Review by cpgr — COMMENTED** [review]
Thanks for this important fix Andy!

File: modules/porous_flow/src/actions/PorousFlowUnsaturated.C (hunk 5 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowUnsaturated.C:164]: Hey @cpgr - thanks for the review.  I'm happy to use str.empty() of course, but i'm curious as to the difference between `== ""` and `.empty()`.  Can you tell me?

Comment (cpgr) [reply modules/porous_flow/src/actions/PorousFlowUnsaturated.C:164]: I think they are probably the same thing, but I just like empty for readability! It certainly wouldn't make any difference to the working of the code, so feel free to ignore.

File: modules/porous_flow/doc/content/source/kernels/PorousFlowHeatVolumetricExpansion.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/kernels/PorousFlowHeatVolumetricExpansion.md:13]: i've now set `use_displaced_mesh = false` in the four Kernels and two Postprocessors of this PR, and i have set
```
  params.suppressParameter<bool>("use_displaced_mesh");
```
so users cannot set it true.  I think it's still worth mentioning the parameter in the documentation, just to alert users to this "peculiarity"

File: modules/porous_flow/doc/content/source/kernels/PorousFlowMassTimeDerivative.md [hunk 9]
```diff
@@ -15,6 +15,11 @@ See [mass lumping](/mass_lumping.md) for details.
...
 
+For mechanically-coupled simulations (where the mesh deforms) the numerical implementation of this Kernel involves the old value of the volumetric strain.  Further information can be found [here](porous_flow/time_derivative.md).  This is assumed to be calculated by a TensorMechanics strain calculator, for instance [ComputeSmallStrain](ComputeSmallStrain.md), with a given `base_name`.  Hence, you should usually employ the same `base_name` for this Kernel as used by your strain calculator.  However, if you wish to include mechanical deformations, but not couple them to the porous flow, simply enter a `base_name` that doesn't exist (eg `base_name = non_existent`) and the volumetric strain won't be included in this Kernel.  (Note: if you want a fully-coupled simulation but accidentally make a typo in your `base_name`, then PorousFlow will assume you don't want to include the volumetric strain!)
+
+!alert note
+Because it contains volumetric strain, this Kernel should `use_displaced_mesh = false`
+
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/kernels/PorousFlowMassTimeDerivative.md:22]: yep

File: modules/porous_flow/doc/content/source/materials/PorousFlowVolumetricStrain.md [hunk 10]
```diff
@@ -1,6 +1,20 @@
...
-!syntax description /Materials/PorousFlowVolumetricStrain
+This Material computes volumetric strain and volumetric strain rate.  It requires a TensorMechanics strain calculator, which will usually be [ComputeSmallStrain](ComputeSmallStrain.md), to compute the Material property `total_strain`.  It then calculates
+
+- `PorousFlow_total_volumetric_strain_qp` which is $\sum_{i}\epsilon_{ii}$, where $\epsilon$ is `total_strain`
+- `PorousFlow_volumetric_strain_rate_qp` which is $\sum_{i}(\epsilon_{ii} - \epsilon_{ii}^{\mathrm{old}})/\mathrm{d}t$.
+
+It also calculates derivatives of these with respect to the PorousFlow variables.
+
+!alert note
+It is possible to supply a `base_name` if your TensorMechanics strain calculator also has a `base_name`.
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowVolumetricStrain.md:11]: Thanks, done

File: modules/porous_flow/doc/content/source/materials/PorousFlowVolumetricStrain.md [hunk 11]
```diff
@@ -1,6 +1,20 @@
...
+
+!alert note
+The derivatives calculated assume Cartesian small strain.  Therefore, if you use a finite-strain calculator, or if you're using other coordinate systems, such as RZ coordinates, the Jacobian won't be exact, which will lead to slightly poorer convergence
+
+This Material provides the time-derivative of the volumetric strain to the [PorousFlowMassVolumetricExpansion](PorousFlowMassVolumetricExpansion.md) and [PorousFlowHeatVolumetricExpansion](PorousFlowHeatVolumetricExpansion.md).  Related information can be found [here](porous_flow/time_derivative.md).
+
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/materials/PorousFlowVolumetricStrain.md:17]: cool

File: modules/porous_flow/doc/content/source/postprocessors/PorousFlowFluidMass.md [hunk 12]
```diff
@@ -16,8 +16,10 @@ in the `phase` input parameter.
 This `Postprocessor` also provides the option to only calculate fluid mass below a
 certain saturation, which can be invoked using the `saturation_threshold` parameter.
 
-!alert note
-The flag `use_displaced_mesh = true` must be used in simulations experiencing solid-mechanical deformation.
+The flag `use_displaced_mesh = false` should be used in all simulations, even those experiencing solid-mechanical deformation.  The reason is that this postprocessor uses the strain calculated by TensorMechanics to automatically compensate for deformed meshes.  Further information may be found [here](porous_flow/time_derivative.md).  Therefore:
+
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/porous_flow/doc/content/source/postprocessors/PorousFlowFluidMass.md:20]: Yep, done that as per above comment

File: modules/porous_flow/src/actions/PorousFlowActionBase.C (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowActionBase.C:456]: thanks, done

File: modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowSinglePhaseBase.C:270]: thanks, done

File: modules/porous_flow/src/actions/PorousFlowUnsaturated.C (hunk 5 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowUnsaturated.C:164]: thanks

File: modules/porous_flow/test/tests/energy_conservation/tests (hunk 6 above)
Comment (WilkAndy) [reply modules/porous_flow/test/tests/energy_conservation/tests:59]: :-)

File: modules/porous_flow/test/tests/mass_conservation/mass13.i (hunk 8 above)
Comment (WilkAndy) [reply modules/porous_flow/test/tests/mass_conservation/mass13.i:1]: thanks, fixed

File: modules/porous_flow/test/tests/mass_conservation/mass12.i (hunk 7 above)
Comment (WilkAndy) [reply modules/porous_flow/test/tests/mass_conservation/mass12.i:1]: thanks, fixed

File: modules/porous_flow/src/actions/PorousFlowFullySaturated.C (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowFullySaturated.C:176]: yep, thanks

**Comment by WilkAndy** [comment]
Hey @loganharbour  - can you review/merge this, please?

**Comment by WilkAndy** [comment]
Requesting review from @loganharbour .   Not much code - a lot of tests/doco changes.  Quite an important change to PorousFlow

**Review by loganharbour — APPROVED** [review]

**Comment by WilkAndy** [comment]
woohoo, thanks @loganharbour - this was an important change to PorousFlow.

**Pull request merged**